File: server/routes/todos.ts
import { Router } from 'express';
import { isAuthenticated } from '../middleware/auth.js';
import { collections } from '../db/collections.js';
import { cache } from '../lib/cache.js';
import { ObjectId } from 'mongodb';

const router = Router();

const getStartOfDay = (date: Date = new Date()): Date => {
  const d = new Date(date);
  d.setUTCHours(0, 0, 0, 0);
  return d;
};

const isOverdue = (scheduledDate: Date): boolean => {
  const today = getStartOfDay();
  const scheduled = getStartOfDay(new Date(scheduledDate));
  return scheduled < today;
};

const POINTS = {
  ON_TIME_COMPLETION: 2,
  LATE_COMPLETION: 1,
  RESCHEDULE_PENALTY: 0,
  EARLY_COMPLETION_BONUS: 1,
};

router.use(isAuthenticated);

router.get('/', async (req, res) => {
  try {
    const userId = req.user!._id;
    const cacheKey = `todos:${userId.toString()}`;

    const cached = cache.get(cacheKey);
    if (cached) {
      res.setHeader('X-Cache', 'HIT');
      return res.json(cached);
    }

    const todos = await (await collections.todos).find(
      { userId },
      {
        sort: { scheduledDate: 1 },
        projection: {
          title: 1,
          subject: 1,
          difficulty: 1,
          questionsTarget: 1,
          completed: 1,
          scheduledDate: 1,
          rescheduledCount: 1,
          originalScheduledDate: 1,
          createdAt: 1,
        }
      }
    ).toArray();

    const todosWithStatus = todos.map((todo: any) => {
      const effectiveScheduledDate = todo.scheduledDate || todo.createdAt;
      const rescheduleCount = typeof todo.rescheduledCount === 'number' ? todo.rescheduledCount : 0;
      return {
        ...todo,
        // Always expose a plain string `id` so the frontend doesn't need to map _id
        id: todo._id.toString(),
        scheduledDate: effectiveScheduledDate,
        rescheduledCount: rescheduleCount,
        isOverdue: !todo.completed && effectiveScheduledDate && isOverdue(effectiveScheduledDate),
      };
    });

    cache.set(cacheKey, todosWithStatus, 120);
    res.setHeader('X-Cache', 'MISS');
    res.json(todosWithStatus);
  } catch (error) {
    console.error('Todos fetch error:', error);
    res.status(500).json({ error: 'Failed to fetch todos' });
  }
});

router.post('/', async (req, res) => {
  try {
    const userId = req.user!._id;
    const { scheduledDate, ...rest } = req.body;

    const scheduled = scheduledDate ? new Date(scheduledDate) : getStartOfDay();

    const todoData = {
      ...rest,
      userId,
      completed: false,
      questionsCompleted: 0,
      scheduledDate: scheduled,
      rescheduledCount: 0,
      createdAt: new Date(),
      updatedAt: new Date()
    };

    const result = await (await collections.todos).insertOne(todoData);
    const todo = await (await collections.todos).findOne({ _id: result.insertedId });

    cache.delete(`todos:${userId.toString()}`);

    res.json({
      ...todo,
      id: todo!._id.toString(),
      isOverdue: false,
    });
  } catch (error) {
    console.error('Todo creation error:', error);
    res.status(500).json({ error: 'Failed to create todo', details: (error as any).message });
  }
});

router.post('/reschedule-all-overdue', async (req, res) => {
  try {
    const userId = req.user!._id;
    const { targetDate } = req.body;

    const scheduleTo = targetDate ? getStartOfDay(new Date(targetDate)) : getStartOfDay();
    const today = getStartOfDay();

    if (scheduleTo < today) {
      return res.status(400).json({ error: 'Cannot schedule tasks in the past' });
    }

    const allTodos = await (await collections.todos).find({ userId, completed: false }).toArray();

    const overdueTodos = allTodos.filter((todo: any) => {
      if (!todo.scheduledDate) return false;
      return isOverdue(todo.scheduledDate);
    });

    if (overdueTodos.length === 0) {
      return res.json({ message: 'No overdue tasks to reschedule', count: 0 });
    }

    const overdueIds = overdueTodos.map(todo => todo._id);
    const result = await (await collections.todos).updateMany(
      { _id: { $in: overdueIds } },
      {
        $set: { scheduledDate: scheduleTo, updatedAt: new Date() },
        $inc: { rescheduledCount: 1 },
      }
    );

    const updatedCount = result.modifiedCount;

    cache.delete(`todos:${userId.toString()}`);

    res.json({
      success: true,
      message: `${updatedCount} task${updatedCount > 1 ? 's' : ''} rescheduled`,
      count: updatedCount,
    });
  } catch (error) {
    console.error('Bulk reschedule error:', error);
    res.status(500).json({ error: 'Failed to reschedule tasks' });
  }
});

router.patch('/:id', async (req, res) => {
  try {
    const userId = req.user!._id;

    if (!ObjectId.isValid(req.params.id)) {
      return res.status(400).json({ error: 'Invalid task ID' });
    }

    const todoId = new ObjectId(req.params.id);

    const existingTodo = await (await collections.todos).findOne({ _id: todoId, userId });

    if (!existingTodo) {
      return res.status(404).json({ error: 'Todo not found' });
    }

    const updateMap: any = { ...req.body, updatedAt: new Date() };
    delete updateMap._id;
    delete updateMap.userId;

    await (await collections.todos).updateOne(
      { _id: todoId },
      { $set: updateMap }
    );

    const updatedTodo = await (await collections.todos).findOne({ _id: todoId });

    let pointsToIncrement = 0;
    let pointsToAward = 0;

    if (req.body.completed && !existingTodo.completed) {
      pointsToAward = 0.5;

      const today = getStartOfDay();
      const scheduledDate = existingTodo.scheduledDate ? getStartOfDay(new Date(existingTodo.scheduledDate)) : today;
      const originalScheduledDate = existingTodo.originalScheduledDate ? getStartOfDay(new Date(existingTodo.originalScheduledDate)) : null;

      if (scheduledDate.getTime() === today.getTime()) {
        if (!originalScheduledDate || originalScheduledDate.getTime() === today.getTime()) {
          pointsToAward = 1;
        } else {
          pointsToAward = 0.5;
        }
      } else if (scheduledDate < today) {
        pointsToAward = 0.5;
      }

      pointsToIncrement = pointsToAward === 0.5 ? 1 : Math.round(pointsToAward);
      if (pointsToIncrement > 0) {
        await (await collections.users).updateOne(
          { _id: userId },
          {
            $inc: { totalPoints: pointsToIncrement },
            $set: { lastActive: new Date() }
          }
        );
      }
    }

    cache.delete(`todos:${userId.toString()}`);

    const safeRescheduledCount = typeof updatedTodo!.rescheduledCount === 'number' ? updatedTodo!.rescheduledCount : 0;

    res.json({
      ...updatedTodo,
      rescheduledCount: safeRescheduledCount,
      isOverdue: !updatedTodo!.completed && updatedTodo!.scheduledDate && isOverdue(updatedTodo!.scheduledDate),
      pointsAwarded: req.body.completed && !existingTodo.completed ? pointsToAward : 0,
    });
  } catch (error) {
    console.error('Todo update error:', error);
    res.status(500).json({ error: 'Failed to update todo' });
  }
});

router.patch('/:id/reschedule', async (req, res) => {
  try {
    const userId = req.user!._id;
    const { newDate } = req.body;

    if (!ObjectId.isValid(req.params.id)) {
      return res.status(400).json({ error: 'Invalid task ID' });
    }

    const todoId = new ObjectId(req.params.id);

    if (!newDate) {
      return res.status(400).json({ error: 'New date is required' });
    }

    const existingTodo = await (await collections.todos).findOne({ _id: todoId, userId });

    if (!existingTodo) {
      return res.status(404).json({ error: 'Todo not found' });
    }

    if (existingTodo.completed) {
      return res.status(400).json({ error: 'Cannot reschedule a completed task' });
    }

    const newScheduledDate = getStartOfDay(new Date(newDate));
    const today = getStartOfDay();

    if (newScheduledDate < today) {
      return res.status(400).json({ error: 'Cannot schedule a task in the past' });
    }

    const currentRescheduledCount = typeof existingTodo.rescheduledCount === 'number' ? existingTodo.rescheduledCount : 0;

    const updateData: any = {
      scheduledDate: newScheduledDate,
      rescheduledCount: currentRescheduledCount + 1,
      updatedAt: new Date()
    };

    if (currentRescheduledCount === 0) {
      updateData.originalScheduledDate = existingTodo.scheduledDate;
    }

    await (await collections.todos).updateOne(
      { _id: todoId },
      { $set: updateData }
    );

    const updatedTodo = await (await collections.todos).findOne({ _id: todoId });

    const pointsToCredit = Math.floor(POINTS.ON_TIME_COMPLETION / 2);

    await (await collections.users).updateOne(
      { _id: userId },
      {
        $inc: { totalPoints: pointsToCredit },
        $set: { lastActive: new Date() }
      }
    );

    cache.delete(`todos:${userId.toString()}`);

    const safeRescheduledCount = typeof updatedTodo!.rescheduledCount === 'number' ? updatedTodo!.rescheduledCount : 0;

    res.json({
      ...updatedTodo,
      rescheduledCount: safeRescheduledCount,
      isOverdue: false,
      message: 'Task rescheduled successfully',
      pointsCredited: pointsToCredit,
    });
  } catch (error) {
    console.error('Todo reschedule error:', error);
    res.status(500).json({ error: 'Failed to reschedule todo' });
  }
});

router.post('/:id/reschedule-to-today', async (req, res) => {
  try {
    const userId = req.user!._id;

    if (!ObjectId.isValid(req.params.id)) {
      return res.status(400).json({ error: 'Invalid task ID' });
    }

    const todoId = new ObjectId(req.params.id);

    const existingTodo = await (await collections.todos).findOne({ _id: todoId, userId });

    if (!existingTodo) {
      return res.status(404).json({ error: 'Todo not found' });
    }

    if (existingTodo.completed) {
      return res.status(400).json({ error: 'Cannot reschedule a completed task' });
    }

    const today = getStartOfDay();

    const currentRescheduledCount = typeof existingTodo.rescheduledCount === 'number' ? existingTodo.rescheduledCount : 0;

    const updateData: any = {
      scheduledDate: today,
      rescheduledCount: currentRescheduledCount + 1,
      updatedAt: new Date()
    };

    if (currentRescheduledCount === 0) {
      updateData.originalScheduledDate = existingTodo.scheduledDate;
    }

    await (await collections.todos).updateOne(
      { _id: todoId },
      { $set: updateData }
    );

    const updatedTodo = await (await collections.todos).findOne({ _id: todoId });

    const pointsToCredit = Math.floor(POINTS.ON_TIME_COMPLETION / 2);

    await (await collections.users).updateOne(
      { _id: userId },
      {
        $inc: { totalPoints: pointsToCredit },
        $set: { lastActive: new Date() }
      }
    );

    cache.delete(`todos:${userId.toString()}`);

    const safeRescheduledCount = typeof updatedTodo!.rescheduledCount === 'number' ? updatedTodo!.rescheduledCount : 0;

    res.json({
      ...updatedTodo,
      rescheduledCount: safeRescheduledCount,
      isOverdue: false,
      message: 'Task rescheduled to today',
      pointsCredited: pointsToCredit,
    });
  } catch (error) {
    console.error('Todo reschedule error:', error);
    res.status(500).json({ error: 'Failed to reschedule todo' });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    const userId = req.user!._id;

    if (!ObjectId.isValid(req.params.id)) {
      return res.status(400).json({ error: 'Invalid task ID' });
    }

    const todoId = new ObjectId(req.params.id);

    await (await collections.todos).deleteOne({ _id: todoId, userId });

    cache.delete(`todos:${userId.toString()}`);

    res.json({ success: true });
  } catch (error) {
    console.error('Todo deletion error:', error);
    res.status(500).json({ error: 'Failed to delete todo' });
  }
});

export default router;

File: server/src/routes/todos.rs
use crate::{
    middleware::auth::{is_authenticated, AuthUser},
    models::{Todo, User},
    AppState,
};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    routing::{get, patch, post},
    Extension, Json, Router,
};
use chrono::{DateTime, Utc};
use futures::stream::TryStreamExt;
use mongodb::bson::{doc, oid::ObjectId};
use serde::{Deserialize, Serialize};
use std::{str::FromStr, sync::Arc};

pub fn router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(get_todos).post(create_todo))
        .route("/reschedule-all-overdue", post(reschedule_all_overdue))
        .route("/:id", patch(update_todo).delete(delete_todo))
        .route("/:id/reschedule", patch(reschedule_todo))
        .route("/:id/reschedule-to-today", post(reschedule_to_today))
        .route_layer(axum::middleware::from_fn_with_state(
            state,
            is_authenticated,
        ))
}

fn get_start_of_day(date: DateTime<Utc>) -> DateTime<Utc> {
    date.date_naive().and_hms_opt(0, 0, 0).unwrap().and_utc()
}

fn is_overdue(scheduled_date: DateTime<Utc>) -> bool {
    let today = get_start_of_day(Utc::now());
    let scheduled = get_start_of_day(scheduled_date);
    scheduled < today
}

const POINTS_ON_TIME_COMPLETION: f64 = 2.0;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TodoResponse {
    #[serde(flatten)]
    pub todo: Todo,
    pub id: String,
    pub is_overdue: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub points_awarded: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub points_credited: Option<f64>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateTodoPayload {
    pub title: String,
    pub subject: String,
    pub difficulty: String,
    pub questions_target: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scheduled_date: Option<DateTime<Utc>>,
}

async fn get_todos(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
) -> Result<Json<Vec<TodoResponse>>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todos_coll = state.db.collection::<Todo>("todos");

    let mut cursor = todos_coll
        .find(doc! { "userId": user_id })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let mut todos = Vec::new();

    while let Some(todo) = cursor.try_next().await.unwrap_or(None) {
        let is_overdue = !todo.completed && is_overdue(todo.scheduled_date);
        let id = todo.id.unwrap().to_string();

        todos.push(TodoResponse {
            todo,
            id,
            is_overdue,
            points_awarded: None,
            points_credited: None,
        });
    }

    Ok(Json(todos))
}

async fn create_todo(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Json(payload): Json<CreateTodoPayload>,
) -> Result<Json<TodoResponse>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todos_coll = state.db.collection::<Todo>("todos");

    let scheduled_date = payload
        .scheduled_date
        .unwrap_or_else(|| get_start_of_day(Utc::now()));

    let new_todo = Todo {
        id: None,
        user_id,
        title: payload.title,
        subject: payload.subject,
        difficulty: payload.difficulty,
        questions_target: payload.questions_target,
        questions_completed: 0,
        completed: false,
        scheduled_date,
        rescheduled_count: 0,
        original_scheduled_date: None,
        created_at: Utc::now(),
        updated_at: Utc::now(),
        completed_at: None,
    };

    let result = todos_coll
        .insert_one(new_todo.clone())
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let mut inserted_todo = new_todo;
    inserted_todo.id = result.inserted_id.as_object_id();
    let id = inserted_todo.id.unwrap().to_string();

    Ok(Json(TodoResponse {
        todo: inserted_todo,
        id,
        is_overdue: false,
        points_awarded: None,
        points_credited: None,
    }))
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateTodoPayload {
    pub title: Option<String>,
    pub subject: Option<String>,
    pub difficulty: Option<String>,
    pub questions_target: Option<i32>,
    pub questions_completed: Option<i32>,
    pub completed: Option<bool>,
}

async fn update_todo(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
    Json(payload): Json<UpdateTodoPayload>,
) -> Result<Json<TodoResponse>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todo_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let todos_coll = state.db.collection::<Todo>("todos");
    let filter = doc! { "_id": todo_id, "userId": user_id };

    let existing_todo = todos_coll
        .find_one(filter.clone())
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "Todo not found".into()))?;

    let mut update_doc = mongodb::bson::Document::new();

    if let Some(c) = payload.completed {
        update_doc.insert("completed", c);
    }
    if let Some(t) = payload.title {
        update_doc.insert("title", t);
    }
    if let Some(s) = payload.subject {
        update_doc.insert("subject", s);
    }
    if let Some(d) = payload.difficulty {
        update_doc.insert("difficulty", d);
    }
    if let Some(qt) = payload.questions_target {
        update_doc.insert("questionsTarget", qt);
    }
    if let Some(qc) = payload.questions_completed {
        update_doc.insert("questionsCompleted", qc);
    }

    update_doc.insert("updatedAt", Utc::now().to_rfc3339());

    if update_doc.is_empty() {
        return Err((StatusCode::BAD_REQUEST, "No valid update fields".into()));
    }

    todos_coll
        .update_one(filter.clone(), doc! { "$set": update_doc })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let updated_todo = todos_coll
        .find_one(filter)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .unwrap();

    let mut points_to_award: f64 = 0.0;

    if payload.completed == Some(true) && !existing_todo.completed {
        points_to_award = 0.5;
        let today = get_start_of_day(Utc::now());
        let scheduled_date = get_start_of_day(existing_todo.scheduled_date);

        let original_scheduled_date = existing_todo
            .original_scheduled_date
            .map(get_start_of_day);

        if scheduled_date == today {
            if original_scheduled_date.is_none() || original_scheduled_date == Some(today) {
                points_to_award = 1.0;
            } else {
                points_to_award = 0.5;
            }
        } else if scheduled_date < today {
            points_to_award = 0.5;
        }

        let points_to_increment = if points_to_award == 0.5 {
            1
        } else {
            points_to_award.round() as i32
        };

        if points_to_increment > 0 {
            let users_coll = state.db.collection::<User>("users");
            let _ = users_coll
                .update_one(
                    doc! { "_id": user_id },
                    doc! {
                        "$inc": { "totalPoints": points_to_increment },
                        "$set": { "lastActive": Utc::now().to_rfc3339() }
                    },
                )
                .await;
        }
    }

    let id = updated_todo.id.unwrap().to_string();
    let is_overdue_flag = !updated_todo.completed && is_overdue(updated_todo.scheduled_date);

    Ok(Json(TodoResponse {
        todo: updated_todo,
        id,
        is_overdue: is_overdue_flag,
        points_awarded: if payload.completed == Some(true) && !existing_todo.completed {
            Some(points_to_award)
        } else {
            None
        },
        points_credited: None,
    }))
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReschedulePayload {
    pub new_date: DateTime<Utc>,
}

async fn reschedule_todo(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
    Json(payload): Json<ReschedulePayload>,
) -> Result<Json<TodoResponse>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todo_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let todos_coll = state.db.collection::<Todo>("todos");
    let filter = doc! { "_id": todo_id, "userId": user_id };

    let existing_todo = todos_coll
        .find_one(filter.clone())
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "Todo not found".into()))?;

    if existing_todo.completed {
        return Err((
            StatusCode::BAD_REQUEST,
            "Cannot reschedule a completed task".into(),
        ));
    }

    let new_scheduled_date = get_start_of_day(payload.new_date);
    let today = get_start_of_day(Utc::now());

    if new_scheduled_date < today {
        return Err((
            StatusCode::BAD_REQUEST,
            "Cannot schedule a task in the past".into(),
        ));
    }

    let mut update_doc = doc! {
        "scheduledDate": new_scheduled_date.to_rfc3339(),
        "rescheduledCount": existing_todo.rescheduled_count + 1,
        "updatedAt": Utc::now().to_rfc3339()
    };

    if existing_todo.rescheduled_count == 0 {
        update_doc.insert(
            "originalScheduledDate",
            existing_todo.scheduled_date.to_rfc3339(),
        );
    }

    todos_coll
        .update_one(filter.clone(), doc! { "$set": update_doc })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let updated_todo = todos_coll.find_one(filter).await.unwrap().unwrap();
    let points_to_credit = (POINTS_ON_TIME_COMPLETION / 2.0).floor() as i32;

    let users_coll = state.db.collection::<User>("users");
    let _ = users_coll
        .update_one(
            doc! { "_id": user_id },
            doc! {
                "$inc": { "totalPoints": points_to_credit },
                "$set": { "lastActive": Utc::now().to_rfc3339() }
            },
        )
        .await;

    let id = updated_todo.id.unwrap().to_string();

    Ok(Json(TodoResponse {
        todo: updated_todo,
        id,
        is_overdue: false,
        points_awarded: None,
        points_credited: Some(points_to_credit as f64),
    }))
}

async fn reschedule_to_today(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
) -> Result<Json<TodoResponse>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todo_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let todos_coll = state.db.collection::<Todo>("todos");
    let filter = doc! { "_id": todo_id, "userId": user_id };

    let existing_todo = todos_coll
        .find_one(filter.clone())
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "Todo not found".into()))?;

    if existing_todo.completed {
        return Err((
            StatusCode::BAD_REQUEST,
            "Cannot reschedule a completed task".into(),
        ));
    }

    let today = get_start_of_day(Utc::now());

    let mut update_doc = doc! {
        "scheduledDate": today.to_rfc3339(),
        "rescheduledCount": existing_todo.rescheduled_count + 1,
        "updatedAt": Utc::now().to_rfc3339()
    };

    if existing_todo.rescheduled_count == 0 {
        update_doc.insert(
            "originalScheduledDate",
            existing_todo.scheduled_date.to_rfc3339(),
        );
    }

    todos_coll
        .update_one(filter.clone(), doc! { "$set": update_doc })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let updated_todo = todos_coll.find_one(filter).await.unwrap().unwrap();
    let points_to_credit = (POINTS_ON_TIME_COMPLETION / 2.0).floor() as i32;

    let users_coll = state.db.collection::<User>("users");
    let _ = users_coll
        .update_one(
            doc! { "_id": user_id },
            doc! {
                "$inc": { "totalPoints": points_to_credit },
                "$set": { "lastActive": Utc::now().to_rfc3339() }
            },
        )
        .await;

    let id = updated_todo.id.unwrap().to_string();

    Ok(Json(TodoResponse {
        todo: updated_todo,
        id,
        is_overdue: false,
        points_awarded: None,
        points_credited: Some(points_to_credit as f64),
    }))
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RescheduleAllPayload {
    pub target_date: Option<DateTime<Utc>>,
}

async fn reschedule_all_overdue(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Json(payload): Json<RescheduleAllPayload>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();

    let schedule_to = payload
        .target_date
        .map(get_start_of_day)
        .unwrap_or_else(|| get_start_of_day(Utc::now()));
    let today = get_start_of_day(Utc::now());

    if schedule_to < today {
        return Err((
            StatusCode::BAD_REQUEST,
            "Cannot schedule tasks in the past".into(),
        ));
    }

    let todos_coll = state.db.collection::<Todo>("todos");

    // Fetch all incomplete tasks
    let mut cursor = todos_coll
        .find(doc! { "userId": user_id, "completed": false })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let mut overdue_ids = Vec::new();

    while let Some(todo) = cursor.try_next().await.unwrap_or(None) {
        if is_overdue(todo.scheduled_date) {
            overdue_ids.push(todo.id.unwrap());
        }
    }

    if overdue_ids.is_empty() {
        return Ok(Json(serde_json::json!({
            "message": "No overdue tasks to reschedule",
            "count": 0
        })));
    }

    let result = todos_coll.update_many(
        doc! { "_id": { "$in": &overdue_ids } },
        doc! {
            "$set": { "scheduledDate": schedule_to.to_rfc3339(), "updatedAt": Utc::now().to_rfc3339() },
            "$inc": { "rescheduledCount": 1 }
        }
    ).await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({
        "success": true,
        "message": format!("{} task(s) rescheduled", result.modified_count),
        "count": result.modified_count
    })))
}

async fn delete_todo(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let todo_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let todos_coll = state.db.collection::<Todo>("todos");
    todos_coll
        .delete_one(doc! { "_id": todo_id, "userId": user_id })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "success": true })))
}
